urllib3/poolmanager.py

```diff
@@ -153,44 +148,23 @@ def urlopen(self, method, url, redirect=True, **kw):
         """
         u = parse_url(url)
         conn = self.connection_from_host(u.host, port=u.port, scheme=u.scheme)
-
         kw['assert_same_host'] = False
         kw['redirect'] = False
         if 'headers' not in kw:
             kw['headers'] = self.headers
 
-        if self.proxy is not None and u.scheme == "http":
-            response = conn.urlopen(method, url, **kw)
-        else:
-            response = conn.urlopen(method, u.request_uri, **kw)
-
-        redirect_location = redirect and response.get_redirect_location()
-        if not redirect_location:
-            return response
-
-        # Support relative URLs for redirecting.
-        redirect_location = urljoin(url, redirect_location)
-
-        # RFC 7231, Section 6.4.4
-        if response.status == 303:
-            method = 'GET'
-
-        retries = kw.get('retries')
         if not isinstance(retries, Retry):
             retries = Retry.from_int(retries, redirect=redirect)
 
-        try:
-            retries = retries.increment(method, url, response=response, _pool=conn)
-        except MaxRetryError:
-            if retries.raise_on_redirect:
-                raise
-            return response
-
-        kw['retries'] = retries
-        kw['redirect'] = redirect
+        if self.proxy is not None and u.scheme == "http":
+            response = conn.urlopen(method, url, retries=retries, **kw)
+        else:
+            response = conn.urlopen(method, u.request_uri, retries=retries, **kw)
 
-        log.info("Redirecting %s -> %s", url, redirect_location)
-        return self.urlopen(method, redirect_location, **kw)
+        if redirect and response.get_redirect_location():
+            kw['redirect'] = redirect
+            return handle_redirect(response, self, retries, method, url, pool=conn, **kw)
```

Comment:
Yeah, I'm not thrilled with `handle_redirect`'s signature at the moment. One thing I thought about was making it an instance method on `RequestMethod`, but I'm not sure if that'd be appropriate.

It's also tricky because a few of those are legitimately required arguments - for example, without `retries`, we'd redirect indefinitely.